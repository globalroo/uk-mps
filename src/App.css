.App-header {
  background-color: #222;
  padding: 20px;
  color: white;
  margin-bottom: 20px;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

.ui.space-around.cards {
	justify-content: space-around;
}

.mp-image {
	object-fit: cover;
	object-position: top;
	max-height: 290px;
}

.image {
	background: url("404.jpeg") !important;
	height:290px;
}
