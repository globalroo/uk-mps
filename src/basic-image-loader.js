import React, { Component } from "react";
import PropTypes from "prop-types";

export class BasicImageLoader extends Component {
	constructor(props) {
		super(props);
		this.state = {
			error: false,
			loaded: false
		};
	}
	onLoad = () => {
		this.setState({
			loaded: true
		});
	}
	onError = () => {
		this.setState({
			error: true
		});
	}
	render() {
		const style = this.state.error ? { display: "none", opacity: 0 } : { opacity: (this.state.loaded ? 1 : 0 ) };
		return <img className="mp-image" style={style} src={this.props.src} onLoad={this.onLoad} onError={this.onError} alt={this.props.alt}/>;
	}
}

BasicImageLoader.propTypes = {
	src: PropTypes.string,
	alt: PropTypes.string,
	styles: PropTypes.object
};
