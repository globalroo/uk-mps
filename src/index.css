body {
  margin: 0;
  padding: 0;
  font-family: sans-serif;
  background-color: #ccc
}
.ui.card>.image, .ui.cards>.card>.image {
	flex:1;
}
