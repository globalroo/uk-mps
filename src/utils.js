const RDF_DESCRIPTOR = "http://www.w3.org/1999/02/22-rdf-syntax-ns#type";
const FULLNAME_BY_SURNAME_KEY = "http://example.com/A5EE13ABE03C4D3A8F1A274F57097B6C";
const FULLNAME_BY_FORENAME_KEY = "http://example.com/F31CBD81AD8343898B49DC65743F0BDF";
const MEMBER_IMAGE_KEY = "https://id.parliament.uk/schema/memberHasMemberImage";
const FAMILY_NAME_KEY = "https://id.parliament.uk/schema/personFamilyName";
const GIVEN_NAME_KEY = "https://id.parliament.uk/schema/personGivenName";
const PARLIAMENT_API = "http://lda.data.parliament.uk/members.json";
const LOCALSTORE_KEY = "membersOfParliament";
const PERSON_TYPE = "Person";

export const getThumbnailPortrait = (id = "404") => `images/${id}.jpeg`;
export const getTailOfKey = (key = "") => key.split("/").pop();
export const getCachedMembers = async () => await localStorage.getItem("membersOfParliament");

const getValueFromKeyPair = (keypairArr = []) => {
	const [keypair = { value: undefined }] = keypairArr;
	const { value } = keypair;
	return value;
};

export const getDataFromParliament = async name => {
	const response = await fetch(
		encodeURI(`${PARLIAMENT_API}?fullName=${name}`)
	);
	const jsonResponse = await response.json();
	const { result = { items: [] } } = jsonResponse;
	const { items = [] } = result;
	const [data] = items;
	return data;
};

export const flattenData = async members => {
	const formattedMembers = [];
	for (const [recordIndex, recordData] of Object.entries(members)) {
		const recordType = getValueFromKeyPair(recordData[RDF_DESCRIPTOR]);
		// Only find people - bunch of Nodes in the raw data
		if (recordType && recordType.includes(PERSON_TYPE)) {
			const mpKey = getTailOfKey(recordIndex);
			formattedMembers.push({
				id: getTailOfKey(mpKey),
				image: getThumbnailPortrait(
					getTailOfKey(getValueFromKeyPair(recordData[MEMBER_IMAGE_KEY]))
				),
				fullNameByForename: getValueFromKeyPair(recordData[FULLNAME_BY_FORENAME_KEY]),
				fullNameBySurname: getValueFromKeyPair(recordData[FULLNAME_BY_SURNAME_KEY]),
				familyName: getValueFromKeyPair(recordData[FAMILY_NAME_KEY]),
				givenName: getValueFromKeyPair(recordData[GIVEN_NAME_KEY]),
				data: await getDataFromParliament(getValueFromKeyPair(recordData[FULLNAME_BY_FORENAME_KEY]))
			});
		}
	}
	localStorage.setItem(LOCALSTORE_KEY, JSON.stringify(formattedMembers));
	return formattedMembers;
};
