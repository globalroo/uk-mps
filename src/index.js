import React from "react";
import ReactDOM from "react-dom";
import App from "src/App";
import registerServiceWorker from "src/registerServiceWorker";

import "semantic-ui-css/semantic.min.css";
import "src/index.css";
import members from "data/processed.json";

// Grab unique party names.
// const partyNames = members.map((rec) => rec.partyName);
// const unique = new Set([...partyNames]);

//TODO: Sort members alphabetically, move this to parser.
members.sort((a, b) => {
	if (a.personFamilyName < b.personFamilyName) return -1;
	if (a.personFamilyName > b.personFamilyName) return 1;
	return 0;
});

ReactDOM.render(<App members={members} />, document.getElementById("root"));
registerServiceWorker();
