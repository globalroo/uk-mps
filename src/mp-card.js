import React from "react";
import PropTypes from "prop-types";
import { format } from "date-fns";
import { BasicImageLoader } from "src/basic-image-loader";

const MISSING_IMAGE = "404";

const partyColours = {
	Conservative: "blue",
	Labour: "red",
	ScottishNationalParty: "yellow",
	Independent: "white",
	LiberalDemocrat: "orange",
	DemocraticUnionistParty: "orange",
	PlaidCymru: "lightGreen",
	GreenParty: "green",
	SinnFÃ©in: "darkOrange",
	IndependentUnionist: "lightBlue",
	UlsterUnionistParty: "blue",
	Speaker: "green"
};

export const getDate = (date) => format(date.split("+").shift(), "MMM YYYY");

export const getBestAvailableImage = (availableImage, twitterImage) => {
	let image = availableImage;
	if (image.includes(MISSING_IMAGE)) {
		if (twitterImage) {
			image = twitterImage;
		}
	}
	return image;
};
export const MPCard = ({ mp }) => {
	const index = mp.partyName.split(" ").join("");
	const color = partyColours[index];
	const classes = `${color} card card-overrides`;
	const availableImage = getBestAvailableImage(mp.localImage, mp.twitterImage);
	return (
		<div className={classes}>
			<div className="image image-container">
				<BasicImageLoader src={availableImage} alt={mp.honorificName} />
			</div>
			<div className="content">
				<div className="header">{mp.honorificName}</div>
				<div className="meta">{mp.constituencyGroupName}</div>
				<div className="description">{mp.positionName}</div>
			</div>
			<div className="extra content">
				<span className="right floated">
					<i className="calendar icon" />
					{getDate(mp.constituencyGroupStartDate)}
				</span>
				<span>
					<i className="id badge icon" />
					{mp.partyName}
				</span>
			</div>
		</div>
	);
};

MPCard.propTypes = {
	mp: PropTypes.object
};
