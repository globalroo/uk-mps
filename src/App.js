import React, { Component } from "react";
import PropTypes from "prop-types";
import { MPCard } from "src/mp-card";

// These css imports are breaking testing. checkout babel-jest and these webpack options
// https://facebook.github.io/jest/docs/en/webpack.html
// One of the projects must have these settings already - an ejected one?

import "src/App.css";

class App extends Component {
	render() {
		return (
			<div className="App">
				<header className="App-header">
					<h1 className="App-title">Members of Parliament - getting data ready for memory game</h1>
					<p className="App-intro">
						Back to project:
						<a className="link" href="https://github.com/globalroo/uk-mps">
							uk-mps
						</a>
					</p>
				</header>

				<div className="ui space-around cards">
					{this.props.members.map((mp, ix) => {
						return <MPCard mp={mp} key={`${mp.id}_mp_card_key`} />;
					})}
				</div>

			</div>
		);
	}
}

App.propTypes = {
	members: PropTypes.object
};

export default App;
