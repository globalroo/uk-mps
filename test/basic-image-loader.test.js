import React from "react";
import * as renderer from "react-test-renderer";
import { BasicImageLoader } from "src/basic-image-loader";

describe("BasicImageLoader tests", () => {
	test("BasicImageLoader is rendered as expected", () => {
		const component = renderer.create(<BasicImageLoader src={"test.jpeg"}u alt={"404"}/>);
		expect(component.toJSON()).toMatchSnapshot();
	});
	test("Loaded condition changes state to loaded", () => {
		const component = renderer.create(<BasicImageLoader src={"error"} alt={"404"}/>);
		const instance = component.getInstance();
		instance.onLoad();
		expect(instance.state.loaded).toBe(true);
	});
	test("Error condition changes state to error", () => {
		const component = renderer.create(<BasicImageLoader src={"error"} alt={"404"}/>);
		const instance = component.getInstance();
		instance.onError();
		expect(instance.state.error).toBe(true);
	});
});
