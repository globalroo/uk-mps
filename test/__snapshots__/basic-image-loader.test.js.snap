// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicImageLoader tests BasicImageLoader is rendered as expected 1`] = `
<img
  alt="404"
  className="mp-image"
  onError={[Function]}
  onLoad={[Function]}
  src="test.jpeg"
  style={
    Object {
      "opacity": 0,
    }
  }
/>
`;
